### DISCUSSION TITLE
Particle penetration using Bender et al boundary handling method

### DISCUSSION THREAD

**Discussion by Clairezsh** [InteractiveComputerGraphics/SPlisHSPlasH#129] [Q&A]
Hi, I' m  trying to integrate Bender et al boundary handling method in my system. 

I do same as how it implemented in SPlisHSPlasH library, but the results didn't look great, which the particles  
 all penetrated the boundary and seems don't under the constraint of the boundary. 

I'd be appretiate so much if you have any ideas about what the problem might be!  
And I also want to know how to make sure that I have implemented this method properly in my code? 

Thanks a lottttt!

**Comment by janbender** [comment]
Hi, I moved the question to the discussion page (which is online since 5 minutes ;-)). Can you give me some more info? How do you generate the volume map in your code?

**Reply by Clairezsh** [reply]
Hi, thank you so much for your response!

I was trying to simulate multiple-fluid flows in a tube. The dynamics of multiple-fluid flow is captured by a mixture model, which is from the paper "Multiple-Fluid SPH Simulation Using a Mixture Model[1]". For boundary handling method I use volume map method. I write the volume map generation code in my own case but almost in the same way as SPlisHSPlasH library. In each simulation step, the queried value of boundary volume  was used in the following equation:
![image](https://user-images.githubusercontent.com/59012440/140602847-fd76a742-1f8c-435f-b03c-45a2af301a40.png)

And the results was like the following picture(all particles penetrate the tube):
At first:
![image](https://user-images.githubusercontent.com/59012440/140604200-b9a18a78-6450-4497-84f9-437daa4bff59.png)
Later:
![image](https://user-images.githubusercontent.com/59012440/140604233-d8466eb9-bbff-4521-8fce-d534448c024e.png)

I've checked my code for many times and try to make sure there are no missing codes compared to the boundary handling code in SPlisHSPlasH library. I also try to print the boundary Volume value out :
![image](https://user-images.githubusercontent.com/59012440/140603267-d210cffc-4673-45c4-bc2b-64a942da3aec.png)

My question is how to make sure that I generate the volume map properly, and why the particle penetration happens(Does the mixture model impose special requirements on the boundary handling?).

Thanks again and thanks a lotttt for the incredible SPH library!! By the way, if you need any more information, just let me know!


> _[1](Bo Ren, Chenfeng Li, Xiao Yan, Ming C. Lin, Javier Bonet, and Shi-Min Hu. 2014. Multiple-Fluid SPH Simulation Using a Mixture Model. ACM Trans. Graph. 33, 5, Article 171 (August 2014), 11 pages.)_

**Comment by janbender** [comment]
The formulas look fine. However, it is hard to guess what is going wrong here. Are you sure that the SDF of the volume map has a correct sign? E.g., if you want to keep particles inside a box you have to negate the standard SDF of a box. 

You could also try to visualize the volume map by rendering a colored point encoding the volume (or signed distance) for each node of a regular grid.

**Reply by Clairezsh** [reply]
OK, thanks for your help!
